### PR TITLE
Add globals support

### LINKED ISSUE #3305: Support "external" modules in Parcel 2
Parcel could allow listing of "external" modules which won't be bundled.

### PR DESCRIPTION
**Pull request by 101arrowz**
# ↪️ Pull Request

Adds support for globals using the `alias` in `package.json`.

Based off of #4846 and #4850.

Fixes #3305.

## 💻 Examples

In `package.json`:
```json
{
    "alias": {
        "react": { "global": "React" },
        "myAliasedPackageName": { "fileName": "./myPkg.js" },
        "finalPackage": "finalPackage.js"
    }
}
```
All of the above syntax forms work, for globals and normal aliases respectively. <strike>The pure string syntax works less reliably because it has to guess that it's not an alias based on whether or not the file exists.</strike> (Removed)

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Included links to related issues/PRs

### CODE REVIEW

**Comment by mischnic** [comment]
Why do we need so many variants of specifying a global? Does some other tool have this?

I think just adding `{"global": "React"}` is less confusing (and not ambigious)
```json
{
    "alias": {
        "react": { "global": "React" },
        "lodash": "always-a-filename-alias.js"
    }
}
```

**Comment by 101arrowz** [comment]
I personally like multiple ways of specifying options. For sure, I think having it possible to do just
```json
{
     "jquery": "$"
}
```
would be important because developers would find that more intuitive and less clunky, especially since they're very unlikely to have a file named `$`. I could remove the `["global", "$"]` method (although right now it's just a byproduct of how I'm representing aliases and could just remain undocumented).

If you'd like to make it a rigid structure, I can do that for sure, but it wouldn't be as user-friendly.

**Comment by mischnic** [comment]
> would be important because developers would find that more intuitive and less clunky,

macOS has a case-insensitive filesystem, so `"react": "React"` wouldn't even be stable across OSs
(if it would check `if(does package of that name exist){ usePackage()} else { useGlobal() }`)

**Comment by 101arrowz** [comment]
Fair point, but they could always just use `window.React` if absolutely necessary. I'll remove the string syntax if you insist, though.

**Comment by 101arrowz** [comment]
Any feedback on the implementation?

File: packages/core/integration-tests/test/globals.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    let b = await bundle(
+      path.join(__dirname, '/integration/global-alias/index.js'),
+    );
+
+    let index = await outputFS.readFile(
+      b.getBundles().find(b => b.name.startsWith('index')).filePath,
+      'utf8',
+    );
+
+    assert(/module\.exports\ =\ React/.test(index));
```
Comment (mischnic) [inline packages/core/integration-tests/test/globals.js:16]: Testing this using  `module.exports = React.createElement...` and
```js
let output = await run(b, {React: {createElement(tag){ return "ok"; }}});

assert.equals(output, "ok");
```
would be less brittle.

Comment (101arrowz) [reply packages/core/integration-tests/test/globals.js]: Fixed in 825a3eace291139a66f74e3df8d5a072488fbb97.

**Comment by 101arrowz** [comment]
@mischnic If there's nothing else that needs changes, I'd recommend merging before the next beta. I have a couple projects I'd be able to move from Webpack to Parcel 2 if Parcel 2 supported globals. By the way, the reason I recommended string syntax was for [similarity to Webpack](https://webpack.js.org/configuration/externals/).

**Comment by mischnic** [comment]
The tests are failing

> I think just adding {"global": "React"} is less confusing (and not ambigious)

We still think this is the right path to take (only having the object and not an array). If someone reads the docs for the first time, the first question would probably be: "why are there two ways to achieve the same thing? There has to be *something*"

> recommended string syntax

Makes sense, but that does unfortunately not really work with `alias`

**Comment by 101arrowz** [comment]
I have already removed the string syntax (check the commit history). I have now fixed the failing tests (at least on my machine), thanks for pointing it out.

**Comment by 101arrowz** [comment]
I have already removed the string syntax (check the commit history). I have now fixed the failing tests (at least on my machine), thanks for pointing it out.

**Pull request merged**